// .ignored/fumadocs-ui/dist/provider/base.js
"use client";
import { I18nProvider } from "../contexts/i18n.js";
import { SearchProvider } from "../contexts/search.js";
import { lazy } from "react";
import { jsx } from "react/jsx-runtime";
import { ThemeProvider, useTheme } from "next-themes";
import { DirectionProvider } from "@radix-ui/react-direction";
//#region src/provider/base.tsx
const DefaultSearchDialog = lazy(() => import("../components/dialog/search-default.js"));
function RootProvider({ children, dir = "ltr", theme = {}, search, i18n }) {
	let body = children;
	if (search?.enabled !== false) body = /* @__PURE__ */ jsx(SearchProvider, {
		SearchDialog: DefaultSearchDialog,
		...search,
		children: body
	});
	if (theme?.enabled !== false) body = /* @__PURE__ */ jsx(ThemeProvider, {
		attribute: "class",
		defaultTheme: "system",
		enableSystem: true,
		disableTransitionOnChange: true,
		...theme,
		children: body
	});
	if (i18n) body = /* @__PURE__ */ jsx(I18nProvider, {
		...i18n,
		children: body
	});
	return /* @__PURE__ */ jsx(DirectionProvider, {
		dir,
		children: body
	});
}
//#endregion
export { RootProvider, useTheme };
