// .ignored/fumadocs-ui/dist/contexts/i18n.js
"use client";
import { createContext, useContext, useMemo, useRef } from "react";
import { usePathname, useRouter } from "fumadocs-core/framework";
import { jsx } from "react/jsx-runtime";
//#region src/contexts/i18n.tsx
const defaultTranslations = {
	search: "Search",
	searchNoResult: "No results found",
	toc: "On this page",
	tocNoHeadings: "No Headings",
	lastUpdate: "Last updated on",
	chooseLanguage: "Choose a language",
	nextPage: "Next Page",
	previousPage: "Previous Page",
	chooseTheme: "Theme",
	editOnGithub: "Edit on GitHub"
};
const I18nContext = createContext({ text: { ...defaultTranslations } });
function I18nLabel(props) {
	return useI18n().text[props.label];
}
function useI18n() {
	return useContext(I18nContext);
}
function I18nProvider({ locales = [], locale, onLocaleChange, children, translations }) {
	const router = useRouter();
	const pathname = usePathname();
	const onChange = (value) => {
		if (onLocaleChange) return onLocaleChange(value);
		const segments = pathname.split("/").filter((v) => v.length > 0);
		if (segments[0] !== locale) segments.unshift(value);
		else segments[0] = value;
		router.push(`/${segments.join("/")}`);
	};
	const onChangeRef = useRef(onChange);
	onChangeRef.current = onChange;
	return /* @__PURE__ */ jsx(I18nContext, {
		value: useMemo(() => ({
			locale,
			locales,
			text: {
				...defaultTranslations,
				...translations
			},
			onChange: (v) => onChangeRef.current(v)
		}), [
			locale,
			locales,
			translations
		]),
		children
	});
}
//#endregion
export { I18nLabel, I18nProvider, defaultTranslations, useI18n };
