// .ignored/fumadocs-ui/dist/contexts/search.js
"use client";
import { Suspense, createContext, use, useEffect, useEffectEvent, useMemo, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
//#region src/contexts/search.tsx
const SearchContext = createContext({
	enabled: false,
	open: false,
	hotKey: [],
	setOpenSearch: () => void 0
});
function useSearchContext() {
	return use(SearchContext);
}
function MetaOrControl() {
	const [key, setKey] = useState("⌘");
	useEffect(() => {
		if (window.navigator.userAgent.includes("Windows")) setKey("Ctrl");
	}, []);
	return key;
}
function SearchProvider({ SearchDialog, children, preload = true, options, hotKey = [{
	key: (e) => e.metaKey || e.ctrlKey,
	display: /* @__PURE__ */ jsx(MetaOrControl, {})
}, {
	key: "k",
	display: "K"
}], links }) {
	const [isOpen, setIsOpen] = useState(preload ? false : void 0);
	const onKeyDown = useEffectEvent((e) => {
		if (hotKey.every((v) => typeof v.key === "string" ? e.key === v.key : v.key(e))) {
			setIsOpen((open) => !open);
			e.preventDefault();
		}
	});
	useEffect(() => {
		window.addEventListener("keydown", onKeyDown);
		return () => {
			window.removeEventListener("keydown", onKeyDown);
		};
	}, [hotKey]);
	return /* @__PURE__ */ jsxs(SearchContext, {
		value: useMemo(() => ({
			enabled: true,
			open: isOpen ?? false,
			hotKey,
			setOpenSearch: setIsOpen
		}), [isOpen, hotKey]),
		children: [isOpen !== void 0 && /* @__PURE__ */ jsx(Suspense, {
			fallback: null,
			children: /* @__PURE__ */ jsx(SearchDialog, {
				open: isOpen,
				onOpenChange: setIsOpen,
				links,
				...options
			})
		}), children]
	});
}
/**
* Show children only when search is enabled via React Context
*/
function SearchOnly({ children }) {
	if (useSearchContext().enabled) return children;
}
//#endregion
export { SearchOnly, SearchProvider, useSearchContext };
