// .ignored/fumadocs-ui/dist/components/toc/default.js
"use client";
import { __exportAll } from "../../_virtual/_rolldown/runtime.js";
import { useI18n } from "../../contexts/i18n.js";
import { cn } from "../../utils/cn.js";
import { mergeRefs } from "../../utils/merge-refs.js";
import { useTOCItems } from "./index.js";
import { useCallback, useEffect, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
import * as Primitive from "fumadocs-core/toc";
//#region src/components/toc/default.tsx
var default_exports = /* @__PURE__ */ __exportAll({
	TOCEmpty: () => TOCEmpty,
	TOCItem: () => TOCItem,
	TOCItems: () => TOCItems
});
function TOCItems({ ref, className, ...props }) {
	const containerRef = useRef(null);
	const items = useTOCItems();
	const [computed, setComputed] = useState(null);
	const onCompute = useCallback(() => {
		const container = containerRef.current;
		if (!container) return;
		if (items.length === 0) {
			setComputed(null);
			return;
		}
		const positions = [];
		for (const item of items) {
			const element = container.querySelector(`a[href="${item.url}"]`);
			if (!element) continue;
			const styles = getComputedStyle(element);
			positions.push([element.offsetTop + parseFloat(styles.paddingTop), element.offsetTop + element.clientHeight - parseFloat(styles.paddingBottom)]);
		}
		setComputed({ positions });
	}, [items]);
	useEffect(() => {
		const container = containerRef.current;
		if (!container) return;
		const observer = new ResizeObserver(onCompute);
		observer.observe(container);
		onCompute();
		return () => {
			observer.disconnect();
		};
	}, [onCompute]);
	return /* @__PURE__ */ jsxs("div", {
		className: "relative",
		children: [computed && /* @__PURE__ */ jsx(TocThumb, { computed }), /* @__PURE__ */ jsx("div", {
			ref: mergeRefs(ref, containerRef),
			className: cn("flex flex-col border-s border-fd-foreground/10", className),
			...props
		})]
	});
}
function TocThumb({ computed }) {
	const ref = useRef(null);
	const tocInfo = Primitive.useTOC();
	function calculate(items) {
		const out = {};
		const startIdx = items.findIndex((item) => item.active);
		if (startIdx === -1) return out;
		const endIdx = items.findLastIndex((item) => item.active);
		out["--track-top"] = `${computed.positions[startIdx][0]}px`;
		out["--track-bottom"] = `${computed.positions[endIdx][1]}px`;
		return out;
	}
	Primitive.useTOCListener((items) => {
		const element = ref.current;
		if (!element) return;
		for (const [k, v] of Object.entries(calculate(items))) element.style.setProperty(k, v);
	});
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: "absolute inset-y-0 inset-s-0 bg-fd-primary w-px transition-[clip-path]",
		style: {
			clipPath: `polygon(0 var(--track-top,0), 100% var(--track-top,0), 100% var(--track-bottom,0), 0 var(--track-bottom,0))`,
			...calculate(tocInfo.get())
		}
	});
}
function TOCEmpty() {
	const { text } = useI18n();
	return /* @__PURE__ */ jsx("div", {
		className: "rounded-lg border bg-fd-card p-3 text-xs text-fd-muted-foreground",
		children: text.tocNoHeadings
	});
}
function TOCItem({ item, ...props }) {
	return /* @__PURE__ */ jsx(Primitive.TOCItem, {
		href: item.url,
		...props,
		className: cn("prose py-1.5 text-sm text-fd-muted-foreground scroll-m-4 transition-colors wrap-anywhere first:pt-0 last:pb-0 data-[active=true]:text-fd-primary hover:text-fd-accent-foreground", item.depth <= 2 && "ps-3", item.depth === 3 && "ps-6", item.depth >= 4 && "ps-8", props.className),
		children: item.title
	});
}
//#endregion
export { TOCEmpty, TOCItem, TOCItems, default_exports };
