// .ignored/fumadocs-mdx/dist/config/index.js
import { t as applyMdxPreset } from "../preset-B0H8HalE.js";
import { t as remarkInclude } from "../remark-include-VDoXZSBN.js";
import { z } from "zod";
//#region ../core/dist/source/schema.js
/**
* Zod 4 schema
*/
const metaSchema = z.object({
	title: z.string().optional(),
	pages: z.array(z.string()).optional(),
	description: z.string().optional(),
	root: z.boolean().optional(),
	defaultOpen: z.boolean().optional(),
	collapsible: z.boolean().optional(),
	icon: z.string().optional()
});
/**
* Zod 4 schema
*/
const pageSchema = z.object({
	title: z.string(),
	description: z.string().optional(),
	icon: z.string().optional(),
	full: z.boolean().optional(),
	_openapi: z.looseObject({}).optional()
});
//#endregion
//#region src/config/define.ts
function defineCollections(options) {
	return options;
}
function defineDocs(options) {
	const dir = options.dir ?? "content/docs";
	return {
		type: "docs",
		dir,
		docs: defineCollections({
			type: "doc",
			dir,
			schema: pageSchema,
			...options?.docs
		}),
		meta: defineCollections({
			type: "meta",
			dir,
			schema: metaSchema,
			...options?.meta
		})
	};
}
function defineConfig(config = {}) {
	return config;
}
//#endregion
export { applyMdxPreset, defineCollections, defineConfig, defineDocs, pageSchema as frontmatterSchema, metaSchema, remarkInclude };
