// .ignored/fumadocs-ui/dist/layouts/flux/page/index.js
"use client";
import { I18nLabel, useI18n } from "../../../contexts/i18n.js";
import { cn } from "../../../utils/cn.js";
import { buttonVariants } from "../../../components/ui/button.js";
import { MarkdownCopyButton, ViewOptionsPopover } from "../../shared/page-actions.js";
import { TOC, TOCProvider } from "./slots/toc.js";
import { Footer } from "./slots/footer.js";
import { Breadcrumb } from "./slots/breadcrumb.js";
import { Container } from "./slots/container.js";
import { createContext, use, useEffect, useState } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
import { Edit } from "lucide-react";
//#region src/layouts/flux/page/index.tsx
const PageContext = createContext(null);
function useDocsPage() {
	const context = use(PageContext);
	if (!context) throw new Error("Please use page components under <DocsPage /> (`fumadocs-ui/layouts/flux/page`).");
	return context;
}
function DocsPage({ tableOfContent: { enabled: tocEnabled, single, ...tocProps } = {}, breadcrumb: { enabled: breadcrumbEnabled = true, ...breadcrumb } = {}, footer: { enabled: footerEnabled = true, ...footer } = {}, full = false, toc = [], slots: defaultSlots = {}, children, ...containerProps }) {
	tocEnabled ??= Boolean(toc.length > 0 || tocProps.header || tocProps.footer);
	const slots = {
		breadcrumb: defaultSlots.breadcrumb ?? Breadcrumb,
		footer: defaultSlots.footer ?? Footer,
		container: defaultSlots.container ?? Container,
		toc: defaultSlots.toc ?? {
			provider: TOCProvider,
			main: TOC
		}
	};
	return /* @__PURE__ */ jsx(PageContext, {
		value: {
			props: { full },
			slots
		},
		children: /* @__PURE__ */ jsxs(slots.toc.provider, {
			single,
			toc: tocEnabled ? toc : [],
			children: [tocEnabled && (tocProps.component ?? /* @__PURE__ */ jsx(slots.toc.main, { ...tocProps })), /* @__PURE__ */ jsxs(slots.container, {
				...containerProps,
				children: [
					breadcrumbEnabled && (breadcrumb.component ?? /* @__PURE__ */ jsx(slots.breadcrumb, { ...breadcrumb })),
					children,
					footerEnabled && (footer.component ?? /* @__PURE__ */ jsx(slots.footer, { ...footer }))
				]
			})]
		})
	});
}
function EditOnGitHub(props) {
	return /* @__PURE__ */ jsx("a", {
		target: "_blank",
		rel: "noreferrer noopener",
		...props,
		className: cn(buttonVariants({
			color: "secondary",
			size: "sm"
		}), "gap-1.5 not-prose", props.className),
		children: props.children ?? /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx(Edit, { className: "size-3.5" }), /* @__PURE__ */ jsx(I18nLabel, { label: "editOnGithub" })] })
	});
}
/**
* Add typography styles
*/
function DocsBody({ children, className, ...props }) {
	return /* @__PURE__ */ jsx("div", {
		...props,
		className: cn("prose flex-1", className),
		children
	});
}
function DocsDescription({ children, className, ...props }) {
	if (children === void 0) return null;
	return /* @__PURE__ */ jsx("p", {
		...props,
		className: cn("mb-8 text-lg text-fd-muted-foreground", className),
		children
	});
}
function DocsTitle({ children, className, ...props }) {
	return /* @__PURE__ */ jsx("h1", {
		...props,
		className: cn("text-[1.75em] font-semibold", className),
		children
	});
}
function PageLastUpdate({ date: value, ...props }) {
	const { text } = useI18n();
	const [date, setDate] = useState("");
	useEffect(() => {
		setDate(value.toLocaleDateString());
	}, [value]);
	return /* @__PURE__ */ jsxs("p", {
		...props,
		className: cn("text-sm text-fd-muted-foreground", props.className),
		children: [
			text.lastUpdate,
			" ",
			date
		]
	});
}
//#endregion
export { DocsBody, DocsDescription, DocsPage, DocsTitle, EditOnGitHub, MarkdownCopyButton, Breadcrumb as PageBreadcrumb, Footer as PageFooter, PageLastUpdate, ViewOptionsPopover, useDocsPage };
