// .ignored/fumadocs-mdx/dist/build-mdx-82nyqL4J.js
import { t as __exportAll } from "./rolldown-runtime-CiIaOW0V.js";
import { n as flattenNode, t as remarkInclude } from "./remark-include-VDoXZSBN.js";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import { createProcessor } from "@mdx-js/mdx";
import { valueToEstree } from "estree-util-value-to-estree";
import { removePosition } from "unist-util-remove-position";
import { mdxToMarkdown } from "mdast-util-mdx";
import { defaultHandlers, toMarkdown } from "mdast-util-to-markdown";
//#region ../core/dist/utils-pSG8hSr9.js
function toMdxExport(name, value) {
	return toMdxExportRaw(name, valueToEstree(value));
}
function toMdxExportRaw(name, expression) {
	return {
		type: "mdxjsEsm",
		value: "",
		data: { estree: {
			type: "Program",
			sourceType: "module",
			body: [{
				type: "ExportNamedDeclaration",
				attributes: [],
				specifiers: [],
				declaration: {
					type: "VariableDeclaration",
					kind: "let",
					declarations: [{
						type: "VariableDeclarator",
						id: {
							type: "Identifier",
							name
						},
						init: expression
					}]
				}
			}]
		} }
	};
}
//#endregion
//#region ../core/dist/mdx-plugins/stringifier.js
function defaultStringifier(config = {}) {
	const { filterMdxAttributes, filterElement = (node) => {
		switch (node.type) {
			case "mdxJsxFlowElement":
			case "mdxJsxTextElement":
				switch (node.name) {
					case "File":
					case "TypeTable":
					case "Callout":
					case "Card": return true;
				}
				return "children-only";
		}
		return true;
	}, stringify, ...customExtension } = config;
	function modHandler(handler, ctx) {
		return function(node, parent, state, info) {
			let visibility = filterElement(node);
			if (visibility === false) return "";
			if (stringify) {
				const v = stringify(node, parent, state, info, ctx);
				if (v) return v;
			}
			const extraInfo = node.data?._stringify;
			if (extraInfo) if (extraInfo === "children-only") visibility = "children-only";
			else if ("text" in extraInfo) return extraInfo.text;
			else node = extraInfo.node;
			if (visibility === "children-only") {
				if (!("children" in node)) return "";
				switch (node.type) {
					case "mdxJsxTextElement":
					case "paragraph": return state.containerPhrasing(node, info);
					case "mdxJsxFlowElement": return state.containerFlow(node, info);
					default: return state.containerFlow({
						type: "root",
						children: node.children
					}, info);
				}
			}
			switch (node.type) {
				case "mdxJsxFlowElement":
				case "mdxJsxTextElement": {
					const stringifiedAttributes = [];
					for (const attr of node.attributes) {
						if (attr.type === "mdxJsxExpressionAttribute") continue;
						if (filterMdxAttributes && !filterMdxAttributes(node, attr)) continue;
						const str = typeof attr.value === "string" ? attr.value : attr.value?.value;
						if (!str) continue;
						stringifiedAttributes.push({
							type: "mdxJsxAttribute",
							name: attr.name,
							value: str
						});
					}
					return handler({
						...node,
						attributes: stringifiedAttributes
					}, parent, state, info);
				}
				default: return handler(node, parent, state, info);
			}
		};
	}
	const customToMarkdown = { handlers: { _custom(node, _, state, info) {
		const handlers = state.handlers;
		for (const k in handlers) handlers[k] = modHandler(handlers[k], node.ctx);
		return state.handle(node.root, void 0, state, info);
	} } };
	return function(root, ctx) {
		return toMarkdown({
			type: "_custom",
			root,
			ctx
		}, {
			...this?.data("settings"),
			extensions: [
				mdxToMarkdown(),
				...this?.data("toMarkdownExtensions") ?? [],
				customToMarkdown,
				customExtension
			]
		});
	};
}
//#endregion
//#region ../core/dist/mdx-plugins/remark-llms.js
/**
* generate `llms.txt` for markdown.
*/
function remarkLLMs({ as = "_markdown", headingIds = true, _data = false, mdxAsPlaceholder, ...rest } = {}) {
	const stringifier = defaultStringifier({
		...rest,
		filterElement(node) {
			switch (node.type) {
				case "mdxjsEsm": return false;
				default: return true;
			}
		},
		stringify(node, parent, state, info, ctx) {
			if (mdxAsPlaceholder) switch (node.type) {
				case "mdxJsxFlowElement":
				case "mdxJsxTextElement": if (node.name && mdxAsPlaceholder.includes(node.name)) return placeholder(node, parent, state, info);
			}
			return rest.stringify?.(node, parent, state, info, ctx);
		},
		handlers: {
			heading(node, _p, state, info) {
				const id = node.data?.hProperties?.id;
				const defaultValue = defaultHandlers.heading(node, _p, state, info);
				return headingIds && id ? `${defaultValue} [#${id}]` : defaultValue;
			},
			...rest.handlers
		}
	});
	return (node, file) => {
		const value = stringifier.call(this, node, void 0);
		node.children.unshift(toMdxExport(as, value));
		if (_data) file.data.markdown = value;
	};
}
/**
* Preserve AST data to render the MDX component at runtime, use `renderPlaceholder()` to render the placeholders.
*/
function placeholder(node, _parent, state, info) {
	const attributes = {};
	for (const attr of node.attributes) {
		if (attr.type === "mdxJsxExpressionAttribute") continue;
		attributes[attr.name] = attr.value;
	}
	return `\0${JSON.stringify({
		name: node.name,
		children: state.containerPhrasing(node, info),
		attributes
	})}\0`;
}
//#endregion
//#region src/loaders/mdx/remark-postprocess.ts
/**
* - collect references
* - write frontmatter (auto-title & description)
*/
function remarkPostprocess({ includeProcessedMarkdown = false, includeMDAST = false, extractLinkReferences = false, valueToExport = [] }) {
	return (tree, file) => {
		const frontmatter = file.data.frontmatter ??= {};
		if (!frontmatter.title) visit(tree, "heading", (node) => {
			if (node.depth === 1) {
				frontmatter.title = flattenNode(node);
				return false;
			}
		});
		file.data["mdx-export"] ??= [];
		file.data["mdx-export"].push({
			name: "frontmatter",
			value: frontmatter
		});
		if (extractLinkReferences) {
			const urls = [];
			visit(tree, "link", (node) => {
				urls.push({ href: node.url });
				return "skip";
			});
			file.data["mdx-export"].push({
				name: "extractedReferences",
				value: urls
			});
		}
		if (includeProcessedMarkdown) remarkLLMs.call(this, typeof includeProcessedMarkdown === "object" ? includeProcessedMarkdown : void 0)(tree, file, () => void 0);
		if (includeMDAST) {
			const mdast = JSON.stringify((includeMDAST === true ? {} : includeMDAST).removePosition ? removePosition(structuredClone(tree)) : tree);
			file.data["mdx-export"].push({
				name: "_mdast",
				value: mdast
			});
		}
		for (const { name, value } of file.data["mdx-export"]) tree.children.unshift(getMdastExport(name, value));
		file.data["mdx-export"] = [];
		for (const name of valueToExport) {
			if (!(name in file.data)) continue;
			tree.children.unshift(getMdastExport(name, file.data[name]));
		}
	};
}
/**
* MDX.js first converts javascript (with esm support) into mdast nodes with remark-mdx, then handle the other remark plugins
*
* Therefore, if we want to inject an export, we must convert the object into AST, then add the mdast node
*/
function getMdastExport(name, value) {
	return {
		type: "mdxjsEsm",
		value: "",
		data: { estree: {
			type: "Program",
			sourceType: "module",
			body: [{
				type: "ExportNamedDeclaration",
				attributes: [],
				specifiers: [],
				source: null,
				declaration: {
					type: "VariableDeclaration",
					kind: "let",
					declarations: [{
						type: "VariableDeclarator",
						id: {
							type: "Identifier",
							name
						},
						init: valueToEstree(value)
					}]
				}
			}]
		} }
	};
}
//#endregion
//#region src/loaders/mdx/build-mdx.ts
var build_mdx_exports = /* @__PURE__ */ __exportAll({ buildMDX: () => buildMDX });
async function buildMDX(core, collection, { filePath, frontmatter, source, _compiler, environment, isDevelopment }) {
	const mdxOptions = await core.getConfig().getMDXOptions(collection, environment);
	function getProcessor(format) {
		const cache = core.cache;
		const key = `build-mdx:${collection?.name ?? "global"}:${format}`;
		let processor = cache.get(key);
		if (!processor) {
			const postprocessOptions = {
				_format: format,
				...collection?.postprocess
			};
			processor = createProcessor({
				outputFormat: "program",
				development: isDevelopment,
				...mdxOptions,
				remarkPlugins: [
					remarkInclude,
					...mdxOptions.remarkPlugins ?? [],
					[remarkPostprocess, postprocessOptions]
				],
				format
			});
			cache.set(key, processor);
		}
		return processor;
	}
	let vfile = new VFile({
		value: source,
		path: filePath,
		cwd: collection?.cwd,
		data: {
			frontmatter,
			_compiler,
			_getProcessor: getProcessor
		}
	});
	if (collection) vfile = await core.transformVFile({
		collection,
		filePath,
		source
	}, vfile);
	return getProcessor(filePath.endsWith(".mdx") ? "mdx" : "md").process(vfile);
}
//#endregion
export { build_mdx_exports as n, buildMDX as t };
