// .ignored/fumadocs-mdx/dist/vite/index.js
import { t as createMdxLoader } from "../mdx-D2vcWdqv.js";
import { t as buildConfig } from "../build-DPafU2lu.js";
import { n as createCore, r as ValidationError, t as _Defaults } from "../core-B0lF760L.js";
import { a as createIntegratedConfigLoader, r as toVite } from "../adapter-DB1MHK9h.js";
import { t as createMetaLoader } from "../meta-B6guI74I.js";
import { t as indexFile } from "../index-file-eVlUbKSO.js";
import { mergeConfig } from "vite";
//#region src/vite/index.ts
async function mdx(config, pluginOptions = {}) {
	const options = applyDefaults(pluginOptions);
	const core = createViteCore(options);
	await core.init({ config: buildConfig(config) });
	const configLoader = createIntegratedConfigLoader(core);
	const mdxLoader = toVite(createMdxLoader(configLoader));
	const metaLoader = toVite(createMetaLoader(configLoader, { json: "json" }));
	return {
		name: "fumadocs-mdx",
		enforce: "pre",
		config(config) {
			if (!options.updateViteConfig) return config;
			return mergeConfig(config, { resolve: {
				noExternal: [
					"fumadocs-core",
					"fumadocs-ui",
					"fumadocs-openapi",
					"@fumadocs/base-ui"
				],
				dedupe: [
					"fumadocs-core",
					"fumadocs-ui",
					"fumadocs-openapi",
					"@fumadocs/base-ui"
				]
			} });
		},
		async buildStart() {
			await core.emit({ write: true });
		},
		async configureServer(server) {
			await core.initServer({ watcher: server.watcher });
		},
		async transform(value, id) {
			try {
				if (metaLoader.filter(id)) return await metaLoader.transform.call(this, value, id);
				if (mdxLoader.filter(id)) return await mdxLoader.transform.call(this, value, id);
			} catch (e) {
				if (e instanceof ValidationError) throw new Error(await e.toStringFormatted(), { cause: e });
				throw e;
			}
		}
	};
}
async function postInstall(pluginOptions = {}) {
	const { loadConfig } = await import("../load-from-file-C1DDNVwF.js").then((n) => n.n);
	const core = createViteCore(applyDefaults(pluginOptions));
	await core.init({ config: loadConfig(core, true) });
	await core.emit({ write: true });
}
function createViteCore({ index, configPath, outDir }) {
	if (index === true) index = {};
	return createCore({
		environment: "vite",
		configPath,
		outDir,
		plugins: [index && indexFile({
			...index,
			target: index.target ?? "vite"
		})]
	});
}
function applyDefaults(options) {
	return {
		updateViteConfig: options.updateViteConfig ?? true,
		index: options.index ?? true,
		configPath: options.configPath ?? _Defaults.configPath,
		outDir: options.outDir ?? _Defaults.outDir
	};
}
//#endregion
export { mdx as default, postInstall };
